File: MS.NET/Fundamentals/Generics/Basics/iterationtest.cs
using System;
using System.Collections;
using System.Collections.Generic;

public class SimpleStack<V> : IEnumerable<V>
{
	class Node
	{
		internal V value;
		internal Node below;

		internal Node(V v, Node n)
		{
			value = v;
			below = n;
		}
	}

	private Node top;

	public void Push(V element)
	{
		top = new Node(element, top);
	}

	public V Pop()
	{
		V element = top.value;
		top = top.below;
	
		return element;
	}

	public bool Empty()
	{
		return top == null;
	}

	public IEnumerator<V> GetEnumerator()
	{
		for(Node n = top; n != null; n = n.below)
			yield return n.value; // return an implementation of IEnumerator from returned values
	}

	IEnumerator IEnumerable.GetEnumerator()
	{
		return GetEnumerator();
	}
}

static class Program
{
	public static void Main()
	{
		int[] numbers = {1, 2, 3, 4, 5, 6};
		Console.WriteLine("All numbers in array");
		foreach(int n in numbers)
			Console.WriteLine(n);
	
		SimpleStack<Interval> intervals = new SimpleStack<Interval>();
		intervals.Push(new Interval(4, 31));
		intervals.Push(new Interval(7, 42));
		intervals.Push(new Interval(6, 23));
		intervals.Push(new Interval(3, 54));
		intervals.Push(new Interval(5, 15));
		Console.WriteLine("All intervals in stack");
		foreach(Interval i in intervals)
			Console.WriteLine(i);


	}
}

File: Java/Fundamentals/Generic/Basics/IterationTest.java
import java.util.Iterator;

class IterationTest{
	
	static class SimpleStack<V> implements Iterable<V>{
		
		private Node top;

		class Node{
			V value;
			Node below;

			Node(V entry){
				value = entry;
				below = top;
			}
		}

		public void push(V item){
			top = new Node(item);
		}

		public V pop(){
			V item = top.value;
			top = top.below;
			return item;
		}

		public boolean empty(){
			return top == null;
		}

		public Iterator<V> iterator(){
			return new Iterator<V>(){
				private Node current = top;

				public boolean hasNext(){
					return current != null;
				}

				public V next(){
					V result = current.value;
					current = current.below;
					return result;
				}
			};
		}
	}

	public static void main(String[] args){
		int[] array = {1, 9, 16, 25, 49};
		System.out.println("All integers in array");
		for(int i : array)
			System.out.println(i);
		SimpleStack<Interval> stack = new SimpleStack<>();
		stack.push(new Interval(5, 31));
		stack.push(new Interval(3, 52));
		stack.push(new Interval(7, 43));
		stack.push(new Interval(6, 14));
		stack.push(new Interval(4, 25));
		System.out.println("All intervals on stack");
		for(Interval i : stack)
			System.out.println(i);
	}
}





